### PR TITLE
Ravendb 4097 first class patch operations

### PR DESCRIPTION
**Pull request by aviv86**

### CODE REVIEW

File: src/Raven.Client/Documents/Session/DocumentSession.Patch.cs [hunk 1]
```diff
@@ -99,12 +100,15 @@ public partial class DocumentSession
...
 
-            _patchInfo = new { DocId = key, Script = $"\"{script}\"", Path = path, Val = val };
+            _patchInfo.Add(new { DocId = key, Script = $"\"{script}\"", Path = path, Val = val });
         }
 
         public override string ToString()
         {
-            return _patchInfo.ToString();
+            var sb = new StringBuilder();
+            foreach (var v in _patchInfo)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/DocumentSession.Patch.cs:109]: This overrides session.ToString().
Use session.Advanced.Defer and you will not need this at all.

Comment (ayende) [reply src/Raven.Client/Documents/Session/DocumentSession.Patch.cs:109]: The idea is that we can see what patches are pending, but this has to be a _separate_ object.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Session/DocumentSession.Patch.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Client/Documents/Session/DocumentSession.Patch.cs:109]: But if this is for debugging then someone can see also the list of deferred commands in the session object?

Comment (ayende) [reply src/Raven.Client/Documents/Session/DocumentSession.Patch.cs:109]: Yes, but not very easily

Comment (ppekrol) [reply src/Raven.Client/Documents/Session/DocumentSession.Patch.cs:109]: Then maybe in ToString we should iterate through the deferred commands and build the debug output based on them? The advantage is that you will see debug information for all pending commands.
Also _patchInfo is not cleared on SaveChanges

**Pull request closed without merging**